### ISSUE TITLE
mmt4d ukernel called with LHS pointer passed also for RHS and OUT

### ISSUE THREAD

**Issue by bjacob** [iree-org/iree#12480]
Debugging the remaining test failures on #12157, namely https://github.com/openxla/iree/pull/12157#issuecomment-1450736576 :

The problem is that `iree_vmvx_mmt4d` is called with incorrect pointers for RHS and OUT: they are getting the exact same pointer value as LHS.

This reproduces already for the 1x1x1 matmul shape case, so here is the corresponding minimal testcase,  `matmul_i8_1x1x1.mlir`:

```mlir
func.func @matmul_i8(%lhs: tensor<1x1xi8>, %rhs: tensor<1x1xi8>, %acc: tensor<1x1xi32>) -> tensor<1x1xi32> {
  %result = linalg.matmul ins(%lhs, %rhs: tensor<1x1xi8>, tensor<1x1xi8>) outs(%acc: tensor<1x1xi32>) -> tensor<1x1xi32>
  return %result: tensor<1x1xi32>
}
```

Here is the command line to reproduce the below print-after-all log (this assumes you've applied #12157 and rebuilt `iree-compile`).

```
tools/iree-compile --iree-hal-target-backends=vmvx --iree-vmvx-enable-microkernels --iree-flow-enable-data-tiling ~/matmul_i8_1x1x1.mlir -o /tmp/a.vmfb --mlir-disable-threading --mlir-print-ir-after-all 2>/tmp/log
```

Here is the print-after-all log:
https://gist.githubusercontent.com/bjacob/841774212c53143361f3f234501b3942/raw/d2617932e4f8e88ce6a45554c5a0981e0e3fc464/gistfile1.txt

Some relevant IR:

```
// -----// IR Dump After LLVMCPULowerExecutableTarget (iree-llvmcpu-lower-executable-target) //----- //
hal.executable.variant public @vmvx_bytecode_fb, target = <"vmvx", "vmvx-bytecode-fb", {ukernels = true}> {
  hal.executable.export public @matmul_i8_dispatch_3_matmul_16x16x16 ordinal(0) layout(#hal.pipeline.layout<push_constants = 0, sets = [<0, bindings = [<0, storage_buffer, ReadOnly>, <1, storage_buffer>]>]>) attributes {translation_info = #iree_codegen.translation_info<VMVXDefault>} {
  ^bb0(%arg0: !hal.device, %arg1: index, %arg2: index, %arg3: index):
    %c1 = arith.constant 1 : index
    %0 = affine.apply affine_map<()[s0] -> (s0 ceildiv 64)>()[%arg1]
    %1 = affine.apply affine_map<()[s0] -> (s0 ceildiv 64)>()[%arg2]
    hal.return %1, %0, %c1 : index, index, index
  }
  builtin.module {
    func.func private @vmvx.mmt4d.i8i8i32(memref<i8>, index, index, memref<i8>, index, index, memref<i32>, index, index, index, index, index, i32, i32, i32, i32) attributes {vm.import.module = "vmvx"}
    func.func @matmul_i8_dispatch_3_matmul_16x16x16() {
      %c1_i32 = arith.constant 1 : i32
      %c0 = arith.constant 0 : index
      %c256 = arith.constant 256 : index
      %c512 = arith.constant 512 : index
      %c16 = arith.constant 16 : index
      %0:2 = vmvx.query_tile_sizes sizes(%c16, %c16) flags(1048576) -> index, index
      %1 = affine.apply affine_map<()[s0] -> (16 ceildiv s0)>()[%0#0]
      %2 = affine.apply affine_map<()[s0] -> (16 ceildiv s0)>()[%0#1]
      %3 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) alignment(64) offset(%c0) flags(ReadOnly) : memref<?x?x?x?xi8>{%1, %2, %0#0, %0#1}
      memref.assume_alignment %3, 64 : memref<?x?x?x?xi8>
      %4:2 = vmvx.query_tile_sizes sizes(%c16, %c16) flags(1114112) -> index, index
      %5 = affine.apply affine_map<()[s0] -> (16 ceildiv s0)>()[%4#0]
      %6 = affine.apply affine_map<()[s0] -> (16 ceildiv s0)>()[%4#1]
      %7 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) alignment(64) offset(%c256) flags(ReadOnly) : memref<?x?x?x?xi8>{%5, %6, %4#0, %4#1}
      memref.assume_alignment %7, 64 : memref<?x?x?x?xi8>
      %8:2 = vmvx.query_tile_sizes sizes(%c16, %c16) flags(1179648) -> index, index
      %9 = affine.apply affine_map<()[s0] -> (16 ceildiv s0)>()[%8#0]
      %10 = affine.apply affine_map<()[s0] -> (16 ceildiv s0)>()[%8#1]
      %11 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) alignment(64) offset(%c512) : memref<?x?x?x?xi32>{%9, %10, %8#0, %8#1}
      memref.assume_alignment %11, 64 : memref<?x?x?x?xi32>
      %workgroup_id_x = hal.interface.workgroup.id[0] : index
      %workgroup_count_x = hal.interface.workgroup.count[0] : index
      %workgroup_id_y = hal.interface.workgroup.id[1] : index
      %workgroup_count_y = hal.interface.workgroup.count[1] : index
      %12 = affine.apply affine_map<()[s0] -> (s0 * 64)>()[%workgroup_id_y]
      %13 = affine.apply affine_map<()[s0] -> (s0 * 64)>()[%workgroup_count_y]
      scf.for %arg0 = %12 to %1 step %13 {
        %14 = affine.min affine_map<(d0)[s0] -> (-d0 + 16 ceildiv s0, 64)>(%arg0)[%0#0]
        %15 = affine.apply affine_map<()[s0] -> (s0 * 64)>()[%workgroup_id_x]
        %16 = affine.apply affine_map<()[s0] -> (s0 * 64)>()[%workgroup_count_x]
        scf.for %arg1 = %15 to %5 step %16 {
          %17 = affine.min affine_map<(d0)[s0] -> (-d0 + 16 ceildiv s0, 64)>(%arg1)[%4#0]
          %subview = memref.subview %3[%arg0, 0, 0, 0] [%14, %2, %0#0, %0#1] [1, 1, 1, 1] : memref<?x?x?x?xi8> to memref<?x?x?x?xi8, strided<[?, ?, ?, 1], offset: ?>>
          %subview_0 = memref.subview %7[%arg1, 0, 0, 0] [%17, %2, %4#0, %0#1] [1, 1, 1, 1] : memref<?x?x?x?xi8> to memref<?x?x?x?xi8, strided<[?, ?, ?, 1], offset: ?>>
          %subview_1 = memref.subview %11[%arg0, %arg1, 0, 0] [%14, %17, %0#0, %4#0] [1, 1, 1, 1] : memref<?x?x?x?xi32> to memref<?x?x?x?xi32, strided<[?, ?, ?, 1], offset: ?>>
          %base_buffer, %offset, %sizes:4, %strides:4 = memref.extract_strided_metadata %subview : memref<?x?x?x?xi8, strided<[?, ?, ?, 1], offset: ?>> -> memref<i8>, index, index, index, index, index, index, index, index, index
          %base_buffer_2, %offset_3, %sizes_4:4, %strides_5:4 = memref.extract_strided_metadata %subview_0 : memref<?x?x?x?xi8, strided<[?, ?, ?, 1], offset: ?>> -> memref<i8>, index, index, index, index, index, index, index, index, index
          %base_buffer_6, %offset_7, %sizes_8:4, %strides_9:4 = memref.extract_strided_metadata %subview_1 : memref<?x?x?x?xi32, strided<[?, ?, ?, 1], offset: ?>> -> memref<i32>, index, index, index, index, index, index, index, index, index
          %18 = arith.index_cast %0#0 : index to i32
          %19 = arith.index_cast %4#0 : index to i32
          %20 = arith.index_cast %0#1 : index to i32
          func.call @vmvx.mmt4d.i8i8i32(%base_buffer, %offset, %strides#0, %base_buffer_2, %offset_3, %strides_5#0, %base_buffer_6, %offset_7, %strides_9#0, %14, %17, %2, %18, %19, %20, %c1_i32) : (memref<i8>, index, index, memref<i8>, index, index, memref<i32>, index, index, index, index, index, i32, i32, i32, i32) -> ()
        }
      }
      return
    }
  }
}
```

There we can see that RHS is in the same binding as LHS but with an offset of 256 (corresponding to the expected 16x16 padded shape) and OUT is in a different binding and with an offset of 512.

So it looks like at least offsets are being ignored, and separately I wonder if bindings are also ignored (as OUT gets the same pointer value in the end, even though it was from a separate binding) or if it's just some kind of aliasing between bindings (as kind of suggested by the offset 512, which is just what it would be if OUT resided in binding 0 after RHS).

**Comment by MaheshRavishankar** [comment]
I am going to describe the issue in as much detail as I can, cause I think this is the last remaining issue, but is uncovering some really bad interaction between IREE and `memref` and `memref.extract_strided_metadata`.

### Problem 

There error seems to happen during the `ResolveBufferDescriptor` pass while using the newer path to micro-kernels using `memref.extract_strided_metadata`. To compare this is the IR before the pass on the failed test.

```
   %3 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) alignment(64) offset(%c0) flags(ReadOnly) : memref<?x?x?x?xi8>{%1, %2, %0#0, %0#1}
   %7 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) alignment(64) offset(%c256) flags(ReadOnly) : memref<?x?x?x?xi8>{%5, %6, %4#0, %4#1}
   %11 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) alignment(64) offset(%c512) : memref<?x?x?x?xi32>{%9, %10, %8#0, %8#1}
   ...
    scf.for %arg0 
      scf.for %arg1
        %subview = memref.subview %3[%arg0, 0, 0, 0] ... : memref<?x?x?x?xi8> to memref<?x?x?x?xi8, strided<[?, ?, ?, 1], offset: ?>>
        %subview_0 = memref.subview %7[%arg1, 0, 0, 0] ... : memref<?x?x?x?xi8> to memref<?x?x?x?xi8, strided<[?, ?, ?, 1], offset: ?>>
        %subview_1 = memref.subview %11[%arg0, %arg1, 0, 0] ... : memref<?x?x?x?xi32> to memref<?x?x?x?xi32, strided<[?, ?, ?, 1], offset: ?>>
        %base_buffer, %offset, %sizes:4, %strides:4 = memref.extract_strided_metadata %subview : memref<?x?x?x?xi8, strided<[?, ?, ?, 1], offset: ?>> -> memref<i8>, index, index, index, index, index, index, index, index, index
        %base_buffer_2, %offset_3, %sizes_4:4, %strides_5:4 = memref.extract_strided_metadata %subview_0 : memref<?x?x?x?xi8, strided<[?, ?, ?, 1], offset: ?>> -> memref<i8>, index, index, index, index, index, index, index, index, index
        %base_buffer_6, %offset_7, %sizes_8:4, %strides_9:4 = memref.extract_strided_metadata %subview_1 : memref<?x?x?x?xi32, strided<[?, ?, ?, 1], offset: ?>> -> memref<i32>, index, index, index, index, index, index, index, index, index
        .... 
        func.call @vmvx.mmt4d.i8i8i32(%base_buffer, %offset, %strides#0, %base_buffer_2, %offset_3, %strides_5#0, %base_buffer_6, %offset_7, %strides_9#0, ....) : (memref<i8>, index, index, memref<i8>, index, index, memref<i32>, index, index, ...)
```

For comparison the path that uses the `vmvx.get_buffer_descriptor` looks like this

```
  %3 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) alignment(64) offset(%c0) flags(ReadOnly) : memref<?x?x?x?xi8>{%1, %2, %0#0, %0#1}
  %7 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) alignment(64) offset(%c256) flags(ReadOnly) : memref<?x?x?x?xi8>{%5, %6, %4#0, %4#1}
  %11 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) alignment(64) offset(%c512) : memref<?x?x?x?xi32>{%9, %10, %8#0, %8#1}
  scf.for %arg0
    scf.for %arg1
      %subview = memref.subview %3[%arg0, 0, 0, 0] ... : memref<?x?x?x?xi8> to memref<?x?x?x?xi8, strided<[?, ?, ?, 1], offset: ?>>
      %base_buffer, %offset, %sizes:4, %strides:4 = vmvx.get_buffer_descriptor %subview : memref<?x?x?x?xi8, strided<[?, ?, ?, 1], offset: ?>> -> !util.buffer, index, index, index, index, index, index, index, index, index
      %subview_0 = memref.subview %7[%arg1, 0, 0, 0] ... : memref<?x?x?x?xi8> to memref<?x?x?x?xi8, strided<[?, ?, ?, 1], offset: ?>>
      %base_buffer_1, %offset_2, %sizes_3:4, %strides_4:4 = vmvx.get_buffer_descriptor %subview_0 : memref<?x?x?x?xi8, strided<[?, ?, ?, 1], offset: ?>> -> !util.buffer, index, index, index, index, index, index, index, index, index
      %subview_5 = memref.subview %11[%arg0, %arg1, 0, 0] ... : memref<?x?x?x?xi32> to memref<?x?x?x?xi32, strided<[?, ?, ?, 1], offset: ?>>
      %base_buffer_6, %offset_7, %sizes_8:4, %strides_9:4 = vmvx.get_buffer_descriptor %subview_5 : memref<?x?x?x?xi32, strided<[?, ?, ?, 1], offset: ?>> -> !util.buffer, index, index, index, index, index, index, index, index, index
      vmvx.mmt4d lhs(%base_buffer offset %offset row_stride %strides#0 : !util.buffer) rhs(%base_buffer_1 offset %offset_2 row_stride %strides_4#0 : !util.buffer) out(%base_buffer_6 offset %offset_7 row_stride %strides_9#0 : !util.buffer) ...
```

These look pretty similar. So far so good. The Interesting part is that after this both the above snippet go through almost exactly the same logic to resolve the `memref.extract_strided_metadata` and `vmvx.get_buffer_descriptors` (it is literally the same templated code). The error is while the first snippet lowers to the following after the pass
```
  %3 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) alignment(64) offset(%c0) flags(ReadOnly) : memref<i8>
  %8 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) alignment(64) offset(%c0) flags(ReadOnly) : memref<i8>
  %13 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) alignment(64) offset(%c0) : memref<i32>
  scf.for %arg0
    scf.for %arg1
      %20 = affine.min affine_map<(d0)[s0] -> (-d0 + 16 ceildiv s0, 64)>(%arg1)[%5#0]
      %21 = affine.apply affine_map<()[s0, s1] -> ((s0 * s1) * (16 ceildiv s0))>()[%0#1, %0#0]
      %22 = affine.apply affine_map<()[s0, s1, s2] -> (s0 + s1 * s2)>()[%c0, %arg0, %21]
      %23 = affine.apply affine_map<()[s0, s1] -> ((s0 * s1) * (16 ceildiv s0))>()[%5#1, %5#0]
      %24 = affine.apply affine_map<()[s0, s1, s2] -> (s0 + s1 * s2)>()[%c0, %arg1, %23]
      %25 = affine.apply affine_map<()[s0, s1] -> (s0 * s1)>()[%10#1, %10#0]
      %26 = affine.apply affine_map<()[s0, s1] -> ((s0 * s1) * (16 ceildiv s0))>()[%10#1, %10#0]
      %27 = affine.apply affine_map<()[s0, s1, s2, s3, s4] -> (s0 + s1 * s2 + s3 * s4)>()[%c0, %arg0, %26, %arg1, %25]
      func.call @vmvx.mmt4d.i8i8i32(%3, %22, %21, %8, %24, %23, %13, %27, %26, ...)
```

the second snippet lowers to
```
  scf.for %arg0
    scf.for %arg1
      %17 = affine.min affine_map<(d0)[s0] -> (-d0 + 16 ceildiv s0, 64)>(%arg1)[%4#0]
      %18 = affine.apply affine_map<()[s0, s1] -> ((s0 * s1) * (16 ceildiv s0))>()[%0#1, %0#0]
      %19 = vmvx.get_raw_interface_binding_buffer set(0) binding(0)
      %20 = affine.apply affine_map<()[s0, s1, s2] -> (s0 + s1 * s2)>()[%c0, %arg0, %18]
      %21 = affine.apply affine_map<()[s0, s1] -> ((s0 * s1) * (16 ceildiv s0))>()[%4#1, %4#0]
      %22 = vmvx.get_raw_interface_binding_buffer set(0) binding(0)
      %23 = affine.apply affine_map<()[s0, s1, s2] -> (s0 + s1 * s2)>()[%c256, %arg1, %21]
      %24 = affine.apply affine_map<()[s0, s1] -> (s0 * s1)>()[%8#1, %8#0]
      %25 = affine.apply affine_map<()[s0, s1] -> ((s0 * s1) * (16 ceildiv s0))>()[%8#1, %8#0]
      %26 = vmvx.get_raw_interface_binding_buffer set(0) binding(1)
      %27 = affine.apply affine_map<()[s0, s1, s2, s3, s4] -> (s0 + s1 * s2 + s3 * s4)>()[%c128, %arg0, %25, %arg1, %24]
      vmvx.mmt4d lhs(%19 offset %20 row_stride %18 : !util.buffer) rhs(%22 offset %23 row_stride %21 : !util.buffer) out(%26 offset %27 row_stride %25 : !util.buffer) ...
```

The difference is 
```
 %24 = affine.apply affine_map<()[s0, s1, s2] -> (s0 + s1 * s2)>()[%c0, %arg1, %23]
 %27 = affine.apply affine_map<()[s0, s1, s2, s3, s4] -> (s0 + s1 * s2 + s3 * s4)>()[%c0, %arg0, %26, %arg1, %25]
```
vs
```
%23 = affine.apply affine_map<()[s0, s1, s2] -> (s0 + s1 * s2)>()[%c256, %arg1, %21]
%27 = affine.apply affine_map<()[s0, s1, s2, s3, s4] -> (s0 + s1 * s2 + s3 * s4)>()[%c128, %arg0, %25, %arg1, %24]
```

So the offset is messed up. Note that in both the cases the base pointer used for the lhs and rhs are the same.

### Triage

Starting with this IR again
```
   %3 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) alignment(64) offset(%c0) flags(ReadOnly) : memref<?x?x?x?xi8>{%1, %2, %0#0, %0#1}
   %7 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) alignment(64) offset(%c256) flags(ReadOnly) : memref<?x?x?x?xi8>{%5, %6, %4#0, %4#1}
   %11 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) alignment(64) offset(%c512) : memref<?x?x?x?xi32>{%9, %10, %8#0, %8#1}
   ...
    scf.for %arg0 
      scf.for %arg1
        %subview = memref.subview %3[%arg0, 0, 0, 0] ... : memref<?x?x?x?xi8> to memref<?x?x?x?xi8, strided<[?, ?, ?, 1], offset: ?>>
        %subview_0 = memref.subview %7[%arg1, 0, 0, 0] ... : memref<?x?x?x?xi8> to memref<?x?x?x?xi8, strided<[?, ?, ?, 1], offset: ?>>
        %subview_1 = memref.subview %11[%arg0, %arg1, 0, 0] ... : memref<?x?x?x?xi32> to memref<?x?x?x?xi32, strided<[?, ?, ?, 1], offset: ?>>
        %base_buffer, %offset, %sizes:4, %strides:4 = memref.extract_strided_metadata %subview : memref<?x?x?x?xi8, strided<[?, ?, ?, 1], offset: ?>> -> memref<i8>, index, index, index, index, index, index, index, index, index
        %base_buffer_2, %offset_3, %sizes_4:4, %strides_5:4 = memref.extract_strided_metadata %subview_0 : memref<?x?x?x?xi8, strided<[?, ?, ?, 1], offset: ?>> -> memref<i8>, index, index, index, index, index, index, index, index, index
        %base_buffer_6, %offset_7, %sizes_8:4, %strides_9:4 = memref.extract_strided_metadata %subview_1 : memref<?x?x?x?xi32, strided<[?, ?, ?, 1], offset: ?>> -> memref<i32>, index, index, index, index, index, index, index, index, index
        .... 
        func.call @vmvx.mmt4d.i8i8i32(%base_buffer, %offset, %strides#0, %base_buffer_2, %offset_3, %strides_5#0, %base_buffer_6, %offset_7, %strides_9#0, ....) : (memref<i8>, index, index, memref<i8>, index, index, memref<i32>, index, index, ...)
```

looking at the debug output the first pattern that triggers is the handling of the

```
%11 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) alignment(64) offset(%c512) : memref<?x?x?x?xi32>{%9, %10, %8#0, %8#1}
%subview_1 = memref.subview %11[%arg0, %arg1, 0, 0] ... : memref<?x?x?x?xi32> to memref<?x?x?x?xi32, strided<[?, ?, ?, 1], offset: ?>>
%base_buffer_6, %offset_7, %sizes_8:4, %strides_9:4 = memref.extract_strided_metadata %subview_1 : memref<?x?x?x?xi32, strided<[?, ?, ?, 1], offset: ?>> -> memref<i32>, index, index, index, index, index, index, index, index, index
func.call @vmvx.mmt4d.i8i8i32(..., %base_buffer_6, %offset_7, %strides_9#0, ...)
```

which resolves to

```
%11 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) alignment(64) offset(%c512) : memref<?x?x?x?xi32>{%9, %10, %8#0, %8#1}
%base_buffer_6, %offset_7, %sizes_8:4, %strides_9:4 = memref.extract_strided_metadata %11 : memref<?x?x?x?xi32> -> memref<i32>, index, index, index, index, index, index, index, index, index
%23 = affine.apply affine_map<()[s0, s1, s2, s3, s4] -> (s0 + s1 * s2 + s3 * s4)>()[%offset_7, %arg0, %strides_9#0, %arg1, %strides_9#1]
func.call @vmvx.mmt4d.i8i8i32(..., %base_buffer_6, %23, %strides_9#0, ...)

```

This is fine. (Even the `vmvx.get_buffer_descriptor` path resovles the same way cause it is actually the same exact code). The error happens since the `memref.extract_strided_metadata` has a folder that looks at the source type, i.e. type of `%11` which is `memref<?x?x?x?xi32>` and folds `%offset_7` to `%c0`. So the above snippet becomes

```
%11 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) alignment(64) offset(%c512) : memref<?x?x?x?xi32>{%9, %10, %8#0, %8#1}
%base_buffer_6, %offset_7, %sizes_8:4, %strides_9:4 = memref.extract_strided_metadata %11 : memref<?x?x?x?xi32> -> memref<i32>, index, index, index, index, index, index, index, index, index
%23 = affine.apply affine_map<()[s0, s1, s2, s3, s4] -> (s0 + s1 * s2 + s3 * s4)>()[%c0, %arg0, %strides_9#0, %arg1, %strides_9#1]
func.call @vmvx.mmt4d.i8i8i32(..., %base_buffer_6, %23, %strides_9#0, ...)
```

This folder is the error, but the root cause is that the modeling of `hal.interface.binding.subspan` and how `memref`s are interpreted upstream seem to incompatible. 

Lets start with this interpretation of the above snippet.

### Solution 1.

1) `hal.interface.binding.subspan set(0) binding(1) offset (%c0)` is basically `char *binding`.
2) `%11` is a `memref` of size `%9 * %10 * %8#0 * %8#1` starting at  base pointer value being `binding + 512`. In other words, the memory region `[binding + 512, binding + 512 + %9 * %10 * %8#0 * %8#1 * elem_size)` is to be interpreted as a `memref<?x?x?x?xi8>`.
3) `%base_buffer_6` is `binding + 512`.

To implement that interpretation I can do

```
%11 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) alignment(64) offset(%c512) : memref<?xi8>{...}
%base_buffer = iree_codegen.get_base_pointer %11
```

This fails during conversion to VM since I end up with a `hal.util.subspan` which cannot be lowered during conversion to VM. Fixing this conversion will make everything connect.

### Solution 2

We can declare bankruptcy on using `memref.extract_strided_metadata` and instead graduate `vmvx.get_buffer_descriptor` to `iree_codegen.get_buffer_descriptor`, and instead of using `memref<...>` we just use `util.buffer` as a way to pass pointer value to the micro kernel calls. This is what the current path through using `vmvx.get_buffer_descriptor` does. So instead start with

```
%11 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) alignment(64) offset(%c512) : memref<?x?x?x?xi32>{%9, %10, %8#0, %8#1}
%subview_1 = memref.subview %11[%arg0, %arg1, 0, 0] ... : memref<?x?x?x?xi32> to memref<?x?x?x?xi32, strided<[?, ?, ?, 1], offset: ?>>
%base_buffer_6, %offset_7, %sizes_8:4, %strides_9:4 = iree_codegen.get_buffer_descriptor %subview_1 : memref<?x?x?x?xi32, strided<[?, ?, ?, 1], offset: ?>> -> !util.buffer, index, index, index, index, index, index, index, index, index
func.call @vmvx.mmt4d.i8i8i32(..., %base_buffer_6, %offset_7, %strides_9#0, ...)
```

which will resolve to
```
%11 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) alignment(64) offset(%c512) : memref<?x?x?x?xi32>{%9, %10, %8#0, %8#1}
%base_buffer_6, %offset_7, %sizes_8:4, %strides_9:4 = iree_codegen.get_buffer_descriptor %11 : memref<?x?x?x?xi32> -> !util.buffer, index, index, index, index, index, index, index, index, index
%23 = affine.apply affine_map<()[s0, s1, s2, s3, s4] -> (s0 + s1 * s2 + s3 * s4)>()[%offset_7, %arg0, %strides_9#0, %arg1, %strides_9#1]
func.call @vmvx.mmt4d.i8i8i32(..., %base_buffer_6, %23, %strides_9#0, ...)
```

which can then resolve to
```
%11 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) alignment(64) offset(%c0) : !util.buffer
%23 = affine.apply affine_map<()[s0, s1, s2, s3, s4] -> (s0 + s1 * s2 + s3 * s4)>()[%c128, %arg0, %strides_9#0, %arg1, %strides_9#1]
func.call @vmvx.mmt4d.i8i8i32(..., %base_buffer_6, %23, %strides_9#0, ...)
```


### Path forward.

@benvanik unless you think we can do something w.r.t lowering of `util.buffer.subspan` in VM conversion (I suspect you dont think that is the way forward), I will just do a sprint to do Solution 2. It has the advantage of this being usable on the LLVM CPU side without having to change upstream cause we only need to teach LLVM converter to handle the `!util.buffer.type` instead of having to deal with overloading/fixing `memref<..>` handling.


FYI @bjacob @stellaraccident @qcolombet

**Comment by MaheshRavishankar** [comment]
FYI for anyone following, Stella and I discussed a path forward here, I am going to give that a shot. Ill update here with what i get.

**Comment by MaheshRavishankar** [comment]
#12510 is the fix that Stella and I discussed. Posting the PR description here

`hal.interface.binding.subspan` ops of the form

```
%subspan = hal.interface.binding.subspan set(0) binding(0) offset(%o0)
    : memref<?x?xf32>{%s0, %s1}
```
are to be interpreted as follows.
- hal.interface.binding.subspan set(0) binding(0) is effectively char *binding.
- A non-zero offset offset(%o0) corresponds to performing binding + %o0.
- The bytes of data from [binding + %o0, binding + %o0 + %s0 * %s1 * <elem-type-byte-width>) can be viewed as a memref<?x?xf32>{%s0, %s1}.

For backends that have a operations which can perform pointer arithmetic (like gep instructions in LLVM), this interpretation can be directly lowered. Backends like VMVX end up using hal.buffer.subspan operations to represent the pointer+offset. This operation does not get lowered to VM dialect directly, but are rather folded away with uses. For cases where we need to get a raw pointer + offset value, these operations dont get folded with uses and fail compilation.

One way to handle this is to convert these operations to represent the entire subspan under memref semantics. So the above operation can be represented as

```
%subspan = hal.interface.binding.subspan set(0) binding(0) offset(%o0)
    : memref<?x?xf32, strided<[?, 1], offset : ?>>{%s0, %s1}
```

giving the subspan effectively subview semantics. This allows use of operations like `memref.extract_strided_metadata` to resolve the offsets, sizes, base pointer usages, etc.

With this the matmul tests pass. But they are other failing tests, probably cause of the new pass added in #12510. Will take a look next week.

**Comment by qcolombet** [comment]
**TL;DR could we stick to using `memref.extract_strided_metadata` while lowering our tensors, at least at the function boundaries, to `memref<..xTy, strided:<[?...], offset: ?>>` from the get go?**

As promised in https://github.com/openxla/iree/pull/12510#issuecomment-1455750217, here are my thoughts on the issue.

At first I was going to say that the default `memref` types are too constrained and not very practical.
Consider the following function:
```
func.func @doSomething(%a : memref<4x4xf32>)
```

This function signature completely forbids any view-like `memref`s by baking in what the input offset and strides should look like (`memref<4x4xf32>` reads `memref<4x4xf32, strided<[4,1], offset: 0>>`).
In other words, we cannot call this function with something like:
```
%alloc = memref.alloc() : memref<32x32xf32>
%subview = memref.subview %alloc2[0, 0][4, 4][1, 1] :
    memref<32x32xf32> to memref<4x4xf32, strided<[32, 1], offset:0>>
```

So I was thinking that maybe the default `memref` type (i.e., when the `strided` information is not provided) should be `strided<[?,?], offset:?>`, that would solve our folding problem but that's not the end of the story.

Now stepping back a little bit, I believe we create the problem for ourselves and although the default `memref` type may not be super friendly, it is not really at fault.
Indeed, our input programs are likely to be in the `tensor` form, therefore, we create this problem when we lower it to `memref`. Instead of using the default `memref` type, we should probably resort on using a variant with `strided<[?,?], offset:?>`, unless we know the strides and offset are always going to match what the default expect.

Put differently, I think that @MaheshRavishankar idea of using non-default strides and offset to describe the `memref` coming from subspan is correct, but I believe it should be applied globally during lowering instead of as a fix-up pass.

**Comment by MaheshRavishankar** [comment]
I think the issue comes down to whether the final lowering supports something like a GEP or not. Going back to this 

> ### Solution 1.
> 1. `hal.interface.binding.subspan set(0) binding(1) offset (%c0)` is basically `char *binding`.
> 2. `%11` is a `memref` of size `%9 * %10 * %8#0 * %8#1` starting at  base pointer value being `binding + 512`. In other words, the memory region `[binding + 512, binding + 512 + %9 * %10 * %8#0 * %8#1 * elem_size)` is to be interpreted as a `memref<?x?x?x?xi8>`.
> 3. `%base_buffer_6` is `binding + 512`.

if you are backend lowers step 2 as a GEP (like the LLVM backends do), then  the region between `[binding + offset, binding + offset + size)` can be interpreted as a contiguous `memref` without any offset (the GEP did the offset) and everything connects. On VMVX backend there is no GEP-like instruction. So for those cases the `memref` needs to account for the offset for things to connect.

> Put differently, I think that @MaheshRavishankar idea of using non-default strides and offset to describe the `memref` coming from subspan is correct, but I believe it should be applied globally during lowering instead of as a fix-up pass.

Agreed! But I am on borrowed time here. I looked at the LLVM lowering of `memref` types. Immediately see some footguns that I dont want to address right now (ill eventually get to it). So I am trying a localized fixup pass. Ideally during bufferization we use the right `memref` type.

**Comment by stellaraccident** [comment]
Mahesh and I went back and forth on this for a bit, trying to split out what the actual semantics of hal.interface.binding.subspan are and the relative odd lengths that the memref type goes to in order to avoid a dependence on a raw pointer type (my interpretation of the situation). I think my conclusion was that there are certainly improvements that can be made to the memref side, but how we were reasoning about the binding.subspan op were just incomplete and was missing a step in lowering: This manifests on targets that lack a gep.

This seemed like a way out of the situation where it is hard to change the semantics of that op in one step atomically for all backends (i.e. from gep to pure cast semantics). It would be great to find out that we could actually just fix it in one step because this is obviously not ideal.

**Comment by MaheshRavishankar** [comment]
> This seemed like a way out of the situation where it is hard to change the semantics of that op in one step atomically for all backends (i.e. from gep to pure cast semantics). It would be great to find out that we could actually just fix it in one step because this is obviously not ideal.

I could change bufferization to do this right from the get-go, i.e.

```
%0 = hal.interface.binding.subspan set(0) binding(1) offset(%offset) : tensor<?x?xf32>{%s0, %s1}
```

bufferizes to

```
%0 = hal.interface.binding.subspan set(0) binding(1) offset(%offset) : memref<?x?xf32, strided<[?, 1], offset: ?>>{%s0, %s1}
```

but I am afraid of what footguns will have to address with this change on the LLVM/NVVM backends. This change currently makes the SPIR-V/VMVX backend over to use the right `memref` semantics for `hal.interface.binding.subspan` operations with offsets. I was planning to come back to this to make the above change during bufferization and drop the `FixupSubspanWithOffsets` pass at that time.
